### PR TITLE
Use `f:dropdownDescriptorSelector` for `securityRealm` & `authorizationStrategy`

### PR DESCRIPTION
**Pull request by jglick**
Avoids using `f:descriptorRadioList` as this is no longer necessary avoid excessive indentation as of #3895, and it is inefficient (loads config block of every possibility even though most are hidden: [JENKINS-20959](https://issues.jenkins.io/browse/JENKINS-20959)); `f:dropdownDescriptorSelector` renders child blocks on demand.

Had to enhance `f:dropdownDescriptorSelector` to show `help.html` where present ([JENKINS-25130](https://issues.jenkins.io/browse/JENKINS-25130)), ~though `f:description` may look better at least for short descriptions.~

Retaining the effect of #3246, pending [JENKINS-20020](https://issues.jenkins.io/browse/JENKINS-20020).

See also #4415 & https://github.com/jenkinsci/workflow-cps-global-lib-plugin/pull/106.

Before: 
sec-before
After: 
sec-after
Demo video:

https://user-images.githubusercontent.com/154109/114934899-86987f00-9e08-11eb-8a22-36ccc9c1f291.mp4

### Proposed changelog entries

* Formatting change in **Configure Security** screen.
* `f:dropdownDescriptorSelector` now honors `help.html` for the selected descriptor.

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins-ci.org/issues/?filter=12146)).

### CODE REVIEW

**Comment by MarkEWaite** [comment]
Does "Allow anonymous read access" disappear from the screen when transitioning from "Logged in users can do anything"

before-logged-in-users

to "anyone can do anything"

before-anyone-can-do-anything

**Comment by jglick** [comment]
> Does "Allow anonymous read access" disappear from the screen when transitioning from "Logged in users can do anything" to "anyone can do anything"

Yes, options defined in a `config.jelly` for a specific `SecurityRealm` or `AuthorizationStrategy` are displayed only when that type is selected.

**Comment by jglick** [comment]
Looking at test failures…

**Comment by jglick** [comment]
First it seems that ATH will need to be tweaked a bit, unsurprisingly:

```
org.openqa.selenium.NoSuchElementException: 
None of the captions exists: Delegate to servlet container
…
	at org.jenkinsci.test.acceptance.po.CapybaraPortingLayerImpl.findCaption(CapybaraPortingLayerImpl.java:460)
	at org.jenkinsci.test.acceptance.po.GlobalSecurityConfig.selectFromRadioGroup(GlobalSecurityConfig.java:87)
	at org.jenkinsci.test.acceptance.po.GlobalSecurityConfig.useRealm(GlobalSecurityConfig.java:52)
	at core.TriggerRemoteBuildsTest.triggerBuildRemotely(TriggerRemoteBuildsTest.java:46)
```

---

The real problem can be seen in both `SecurityRealmTest.submitPossibleWithoutJellyTrace` and `Jenkins41511Test.configRoundTrip`:

```
org.kohsuke.stapler.NoStaplerConstructorException: There's no @DataBoundConstructor on any constructor of class hudson.security.SecurityRealm$None
	at org.kohsuke.stapler.ClassDescriptor.loadConstructorParamNames(ClassDescriptor.java:265)
	at org.kohsuke.stapler.RequestImpl.instantiate(RequestImpl.java:786)
	at org.kohsuke.stapler.RequestImpl.access$200(RequestImpl.java:85)
	at org.kohsuke.stapler.RequestImpl$TypePair.convertJSON(RequestImpl.java:690)
Caused: java.lang.IllegalArgumentException: Failed to instantiate class hudson.security.SecurityRealm from {"stapler-class":"hudson.security.SecurityRealm$None","$class":"hudson.security.SecurityRealm$None"}
	at org.kohsuke.stapler.RequestImpl$TypePair.convertJSON(RequestImpl.java:693)
	at org.kohsuke.stapler.RequestImpl.bindJSON(RequestImpl.java:490)
	at org.kohsuke.stapler.RequestImpl.bindJSON(RequestImpl.java:486)
	at hudson.security.GlobalSecurityConfiguration.configure(GlobalSecurityConfiguration.java:133)
	at hudson.security.GlobalSecurityConfiguration.doConfigure(GlobalSecurityConfiguration.java:117)
```

Somehow https://github.com/jenkinsci/jenkins/blob/f8027c1886eb376ee559f3c47587cb3e91563ee0/core/src/main/java/hudson/security/SecurityRealm.java#L712-L715 and https://github.com/jenkinsci/jenkins/blob/f8027c1886eb376ee559f3c47587cb3e91563ee0/core/src/main/java/hudson/security/AuthorizationStrategy.java#L233-L236 are not being honored.

**Comment by jglick** [comment]
I thought I had already fixed this class of problem in #1936, looking again.

**Comment by jglick** [comment]
I can easily reproduce the ATH failure, but unfortunately I can find no examples of ATH page objects actually picking something out of a `f:dropdownDescriptorSelector`, and I am pretty lost in all the Selenium magic. https://github.com/jenkinsci/acceptance-test-harness/blob/a9027d40be247fbc51f4ea4af7e4bd0bd7f7e6e8/src/main/java/org/jenkinsci/test/acceptance/po/DumbSlave.java#L25 is not encouraging (in that case https://github.com/jenkinsci/jenkins/blob/c60345fb8f3535146c96e36c80af85ee57f41691/core/src/main/resources/hudson/slaves/DumbSlave/configure-entries.jelly#L51 a lower-level Jelly control is in use for whatever reason). It seems that https://github.com/jenkinsci/jenkins/blob/c60345fb8f3535146c96e36c80af85ee57f41691/core/src/main/resources/hudson/security/GlobalSecurityConfiguration/index.groovy#L34 becomes

```html
<select class="setting-input dropdownList" path="/">
…
```

while https://github.com/jenkinsci/jenkins/blob/c60345fb8f3535146c96e36c80af85ee57f41691/core/src/main/resources/hudson/security/GlobalSecurityConfiguration/index.groovy#L39 becomes

```html
<select class="setting-input dropdownList" path="/[1]">
…
```

and https://github.com/jenkinsci/jenkins/blob/c60345fb8f3535146c96e36c80af85ee57f41691/core/src/main/resources/hudson/security/GlobalSecurityConfiguration/index.groovy#L43 becomes

```html
<select class="setting-input dropdownList" path="/[2]">
  <option value="0" selected="true">Plain text</option>
</select>
```

which is not very helpful from the `form-element-path` plugin. (Which BTW would better be baked into core and activated with some system property?) Maybe because https://github.com/jenkinsci/jenkins/blob/c60345fb8f3535146c96e36c80af85ee57f41691/core/src/main/resources/lib/form/dropdownList.jelly#L51-L65 does not put the `field` ~ `name` on the `<select>` itself?

**Comment by jglick** [comment]
```diff
diff --git core/src/main/resources/lib/form/dropdownList.jelly core/src/main/resources/lib/form/dropdownList.jelly
index 71e63400f4..80e5b2d63d 100644
--- core/src/main/resources/lib/form/dropdownList.jelly
+++ core/src/main/resources/lib/form/dropdownList.jelly
@@ -50,6 +50,7 @@ THE SOFTWARE.
             <!-- create drop-down list -->
             <select class="setting-input dropdownList"
                     disabled="${readOnlyMode ? 'true' : null}"
+                    name="${name}"
             >
                 <j:set var="dropdownListMode" value="createSelectField"/>
                 <d:invokeBody/>
```

improves things somewhat in that ATH can then

```diff
diff --git src/main/java/org/jenkinsci/test/acceptance/po/GlobalSecurityConfig.java src/main/java/org/jenkinsci/test/acceptance/po/GlobalSecurityConfig.java
index ec604db4..0bd02408 100644
--- src/main/java/org/jenkinsci/test/acceptance/po/GlobalSecurityConfig.java
+++ src/main/java/org/jenkinsci/test/acceptance/po/GlobalSecurityConfig.java
@@ -49,6 +49,11 @@ public class GlobalSecurityConfig extends ContainerPageObject {
 
     public <T extends SecurityRealm> T useRealm(Class<T> type) {
         maybeCheckUseSecurity();
+        try {
+            control("/securityRealm").selectDropdownMenuAlt(type);
+            return newInstance(type, this, "/securityRealm[1]");
+        } catch (NoSuchElementException x) {
+        }
         return selectFromRadioGroup(type);
     }
 
```

but this just breaks form submission in Jenkins because you then receive e.g.

```json
"securityRealm":   [
  "1",
      {
    "allowsSignup": true,
    "stapler-class": "hudson.security.HudsonPrivateSecurityRealm",
    "$class": "hudson.security.HudsonPrivateSecurityRealm"
  }
]
```

**Comment by jglick** [comment]
```diff
diff --git core/src/main/resources/lib/form/dropdownList.jelly core/src/main/resources/lib/form/dropdownList.jelly
index 71e63400f4..2b1dc52106 100644
--- core/src/main/resources/lib/form/dropdownList.jelly
+++ core/src/main/resources/lib/form/dropdownList.jelly
@@ -50,6 +50,7 @@ THE SOFTWARE.
             <!-- create drop-down list -->
             <select class="setting-input dropdownList"
                     disabled="${readOnlyMode ? 'true' : null}"
+                    name="${name}-select"
             >
                 <j:set var="dropdownListMode" value="createSelectField"/>
                 <d:invokeBody/>
```

plus

```diff
diff --git src/main/java/org/jenkinsci/test/acceptance/plugins/build_timeout/BuildTimeout.java src/main/java/org/jenkinsci/test/acceptance/plugins/build_timeout/BuildTimeout.java
index 3a5b0983..674b2ab6 100644
--- src/main/java/org/jenkinsci/test/acceptance/plugins/build_timeout/BuildTimeout.java
+++ src/main/java/org/jenkinsci/test/acceptance/plugins/build_timeout/BuildTimeout.java
@@ -41,7 +41,7 @@ public class BuildTimeout extends PageAreaImpl {
     private void chooseStrategy(String name) {
         // JENKINS-20164
         if (job.getJenkins().getPlugin("build-timeout").isNewerThan("1.14")) {
-            control("/").select(name);
+            control("/").select(name); // TODO
         } else {
             choose(name);
         }
diff --git src/main/java/org/jenkinsci/test/acceptance/plugins/git/GitScm.java src/main/java/org/jenkinsci/test/acceptance/plugins/git/GitScm.java
index ce4e5c3c..771202a4 100644
--- src/main/java/org/jenkinsci/test/acceptance/plugins/git/GitScm.java
+++ src/main/java/org/jenkinsci/test/acceptance/plugins/git/GitScm.java
@@ -38,7 +38,7 @@ public class GitScm extends Scm {
     private final Control branch = control("branches/name");
     private final Control url = control("userRemoteConfigs/url");
     private final Control tool = control("gitTool");
-    private final Control repositoryBrowser = control("/");
+    private final Control repositoryBrowser = control("/"); // TODO
     private final Control urlRepositoryBrowser = control("browser/repoUrl");
 
     public GitScm(Job job, String path) {
diff --git src/main/java/org/jenkinsci/test/acceptance/plugins/ssh_slaves/SshSlaveLauncher.java src/main/java/org/jenkinsci/test/acceptance/plugins/ssh_slaves/SshSlaveLauncher.java
index 59a2a541..404d3345 100644
--- src/main/java/org/jenkinsci/test/acceptance/plugins/ssh_slaves/SshSlaveLauncher.java
+++ src/main/java/org/jenkinsci/test/acceptance/plugins/ssh_slaves/SshSlaveLauncher.java
@@ -26,7 +26,7 @@ public class SshSlaveLauncher extends ComputerLauncher {
     public final Control retries = control("maxNumRetries");
     public final Control javaPath = control("javaPath");
     public final Control jvmOptions = control("jvmOptions");
-    public final Control hostKeyVerificationStrategy = control("/");
+    public final Control hostKeyVerificationStrategy = control("/"); // TODO
 
     public SshSlaveLauncher(PageObject context, String path) {
         super(context, path);
diff --git src/main/java/org/jenkinsci/test/acceptance/po/DumbSlave.java src/main/java/org/jenkinsci/test/acceptance/po/DumbSlave.java
index d1f1d2d8..9cef8b04 100644
--- src/main/java/org/jenkinsci/test/acceptance/po/DumbSlave.java
+++ src/main/java/org/jenkinsci/test/acceptance/po/DumbSlave.java
@@ -22,7 +22,7 @@ public class DumbSlave extends Slave {
     public final Control executors = control("/numExecutors");
     public final Control remoteFS = control("/remoteFS");
     public final Control labels = control("/labelString");
-    public final Control launchMethod = control("/");   // TODO: this path looks rather buggy to me
+    public final Control launchMethod = control("/launcher-select", /* prior to https://github.com/jenkinsci/jenkins/pull/5417 */ "/");
 
     public DumbSlave(Jenkins j, String name) {
         super(j, name);
diff --git src/main/java/org/jenkinsci/test/acceptance/po/GlobalSecurityConfig.java src/main/java/org/jenkinsci/test/acceptance/po/GlobalSecurityConfig.java
index ec604db4..8644855a 100644
--- src/main/java/org/jenkinsci/test/acceptance/po/GlobalSecurityConfig.java
+++ src/main/java/org/jenkinsci/test/acceptance/po/GlobalSecurityConfig.java
@@ -49,12 +49,24 @@ public class GlobalSecurityConfig extends ContainerPageObject {
 
     public <T extends SecurityRealm> T useRealm(Class<T> type) {
         maybeCheckUseSecurity();
-        return selectFromRadioGroup(type);
+        try {
+            control("/securityRealm-select").selectDropdownMenuAlt(type);
+            return newInstance(type, this, "/securityRealm");
+        } catch (NoSuchElementException x) {
+            // prior to https://github.com/jenkinsci/jenkins/pull/5417
+            return selectFromRadioGroup(type);
+        }
     }
 
     public <T extends AuthorizationStrategy> T useAuthorizationStrategy(Class<T> type) {
         maybeCheckUseSecurity();
-        return selectFromRadioGroup(type);
+        try {
+            control("/authorizationStrategy-select").selectDropdownMenuAlt(type);
+            return newInstance(type, this, "/authorizationStrategy");
+        } catch (NoSuchElementException x) {
+            // prior to https://github.com/jenkinsci/jenkins/pull/5417
+            return selectFromRadioGroup(type);
+        }
     }
 
     private void maybeCheckUseSecurity() {
diff --git src/main/java/org/jenkinsci/test/acceptance/po/LoggedInAuthorizationStrategy.java src/main/java/org/jenkinsci/test/acceptance/po/LoggedInAuthorizationStrategy.java
index ff057bb3..e827b35d 100644
--- src/main/java/org/jenkinsci/test/acceptance/po/LoggedInAuthorizationStrategy.java
+++ src/main/java/org/jenkinsci/test/acceptance/po/LoggedInAuthorizationStrategy.java
@@ -7,7 +7,6 @@ package org.jenkinsci.test.acceptance.po;
  */
 @Describable("Logged-in users can do anything")
 public class LoggedInAuthorizationStrategy extends AuthorizationStrategy {
-    private final Control name = control("/");
     private final Control allowAnonymousRead = control("allowAnonymousRead");
 
     public LoggedInAuthorizationStrategy(GlobalSecurityConfig context, String path) {
```

does somewhat better but still fails in `SshSlavesPluginTest.newSlave` due to some issue with `credentials-dialog-form`.

File: essentials.yml
Comment (jglick) [inline essentials.yml:4]: https://github.com/jenkinsci/acceptance-test-harness/pull/657

**Comment by timja** [comment]
Spacing looks a bit weird because of the help icon 😢 

image

**Comment by jglick** [comment]
Yeah the visual appearance of `help.html` with a dropdown is not great. At least for options implemented in core, I could switch to an `f:description` (perhaps with a help link for more details), but we would still be in the same situation for e.g. `ldap` or `matrix-auth`. Not sure if there is a nicer approach.

**Comment by timja** [comment]
in pipeline we put the symbol name on the same line as the help icon.

i.e.

archiveArtifacts         (?)

Which at least for that use case made it look better,

There was also some thinking on whether we should make the dropdown less wide and put the help button next to it there.

There's some discussion on the PR that changed the behaviour but I'm not sure which pipeline plugin has the snippet generator in it

**Comment by jglick** [comment]
> put the symbol name on the same line as the help icon

Certainly possible. I thought of something like that but was concerned it would look weird to basically have the same text repeated

```
[v] Use SAML 2.0
Use SAML 2.0                  (?)
```

> make the dropdown less wide and put the help button next to it there

Not sure this is feasible due to the way the config block gets lazy-loaded: the help button needs to be rendered inside the box beneath the dropdown. Happy to be proven wrong. Maybe there is some JS trick possible here.

> I'm not sure which pipeline plugin has the snippet generator in it

`workflow-cps-plugin`

**Comment by timja** [comment]
PR was https://github.com/jenkinsci/workflow-cps-plugin/pull/411

Felix did manage to get it working next to the dropdown I believe, should be in the commit history either in the PR or in Liam's fork

**Comment by jglick** [comment]
https://github.com/jenkinsci/workflow-cps-plugin/pull/411#issuecomment-779144806 indeed, but it seems @fqueiruga ultimately abandoned that approach?

**Comment by jglick** [comment]
https://github.com/jenkinsci/workflow-cps-plugin/pull/411/commits/8a9a07f82521862f41a7e8ba484a618f3260b71d vs. https://github.com/jenkinsci/workflow-cps-plugin/pull/411/commits/eabdba1b4b8f4bd657d940afd3588d8c6e8400ce specifically (also CC @bitwiseman).

**Review by oleg-nenashev — DISMISSED** [review]
Tested it locally, looks good for RoleStrategy and other plugins. Have not tested matrix auth though

**Comment by jglick** [comment]
> Have not tested matrix auth

Works fine AFAICT.

File: core/src/main/resources/hudson/security/GlobalSecurityConfiguration/index.groovy
Comment (amuniz) [inline core/src/main/resources/hudson/security/GlobalSecurityConfiguration/index.groovy:39]: @jglick this change breaks a CloudBees proprietary feature (security enforcement from Operations Center to controllers). Can you flag this (internally) as a breaking change please?

Comment (jglick) [reply core/src/main/resources/hudson/security/GlobalSecurityConfiguration/index.groovy:39]: reference for @cloudbees: `OperationsCenterSecurityInfo`

**Comment by daniel-beck** [comment]
> > Have not tested matrix auth
> 
> Works fine AFAICT.

Seems fine in a quick test switching security realms around.

**Review by res0nance — DISMISSED** [review]

**Comment by jglick** [comment]
I think this is still desirable and can certainly deal with the merge conflicts. I need to check the significance of https://github.com/jenkinsci/jenkins/pull/5417#discussion_r635141943. The main task is decide on how best to handle the `f:dropdownDescriptorSelector` change:
1. As here, help button on its own line. Works but is unattractive.
2. Same for compatibility, but start moving `help.jelly` to `f:description` in core and later in plugins defining options for dropdowns. Looks better, but also clutters config display with help text whether you needed it or not.
3. Either of the above, but make this behavior opt-in for `f:dropdownDescriptorSelector` callers, initially enabled only for `GlobalSecurityConfiguration/index.groovy`.
4. Move the button button next to the dropdown (https://github.com/jenkinsci/jenkins/pull/5417#issuecomment-822514985) if possible. Might be beyond my JS/CSS skill level, so help from people with more GUI experience would be great.

**Comment by jglick** [comment]
Note that credentials GUI is not currently affected by this: https://github.com/jenkinsci/credentials-plugin/blob/32810fbc0f2f78ebd5d7fb6d8aa4e778fad1e780/src/main/resources/com/cloudbees/plugins/credentials/CredentialsStoreAction/DomainWrapper/newCredentials.jelly#L50-L51 https://github.com/jenkinsci/credentials-plugin/blob/32810fbc0f2f78ebd5d7fb6d8aa4e778fad1e780/src/main/resources/com/cloudbees/plugins/credentials/CredentialsSelectHelper/WrappedCredentialsStore/credential.jelly#L29-L32

File: core/src/main/resources/lib/form/dropdownDescriptorSelector.jelly
Comment (timja) [inline core/src/main/resources/lib/form/dropdownDescriptorSelector.jelly:70]: This improves the spacing and contrast, top one is after, bottom one is before.

image



```suggestion
              <p class="jenkins-form-label">${%Overview}
                <f:helpLink url="${help}" featureName="${descriptor.displayName}"/>
              </p>
```

Comment (jglick) [reply core/src/main/resources/lib/form/dropdownDescriptorSelector.jelly]: (trying something else)

**Review by timja — COMMENTED** [review]
Looks nice 😄

File: core/src/main/resources/lib/form/dropdownList.jelly
Comment (timja) [inline core/src/main/resources/lib/form/dropdownList.jelly:50]: move to layout.less?

File: core/src/main/resources/lib/form/dropdownList.jelly
Comment (timja) [inline core/src/main/resources/lib/form/dropdownList.jelly:49]: add the style to a css rule?

**Comment by jglick** [comment]
Seems to be working well in Chromium:

https://user-images.githubusercontent.com/154109/160705912-67fb10c9-f3b0-4816-bfc0-5c0364e8f166.mp4

Also did a sanity check in Firefox (not shown).

**Review by NotMyFault — APPROVED** [review]
Looks awesome 👏🏻

**Review by alecharp — APPROVED** [review]
This looks great. 
I tested using a docker container running `jenkins/core-pr-tester` image and it's an improvement for me.

**Review by timja — APPROVED** [review]

**Comment by timja** [comment]
This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Thanks!

**Comment by basil** [comment]
Causes [JENKINS-68748](https://issues.jenkins.io/browse/JENKINS-68748).

**Pull request merged**
